### PR TITLE
Extended MemNN agent added that supports reward-based and forward prediction

### PR DESCRIPTION
**Pull request by ekmb**

### CODE REVIEW

File: examples/train_model.py [hunk 1]
```diff
@@ -123,6 +124,11 @@ def __init__(self, parser):
             build_dict.build_dict(opt)
         # Create model and assign it to the specified task
         self.agent = create_agent(opt)
+        
+        # CUDA with multiprocessing requires 'spawn' start method
+        if opt.get('numthreads', 1) > 1 and opt.get('cuda', False):
+            set_start_method('spawn')
```
Comment (alexholdenmiller) [inline examples/train_model.py]: I think we need to protect this call using `if __name__ == '__main__'`--so this should be in the training loop

File: parlai/tasks/dbll_babi/agents.py [hunk 2]
```diff
@@ -55,8 +55,113 @@ def __init__(self, opt, shared=None):
...
+        self.opt = opt
         super().__init__(opt, shared)
 
+    def setup_data(self, path):
+        if len(self.opt['task'].split(':')) == 4:
+            return self.setup_data_with_feedback(path)
+        else:
+            return super().setup_data(path)
+
+    def setup_data_with_feedback(self, path):
```
Comment (alexholdenmiller) [inline parlai/tasks/dbll_babi/agents.py]: @jaseweston what do you think of this input setup?

File: examples/train_model.py (hunk 1 above)
Comment (alexholdenmiller) [reply examples/train_model.py]: So at the bottom of the train file under the name/main check we should set multiprocessing to spawn if cuda is enabled (we could wrap that in a try/except checking if `import torch; if torch.is_cuda_enabled()` is true as well?)

**Comment by alexholdenmiller** [comment]
if Jason approves than I'm okay for now

File: examples/train_model.py (hunk 1 above)
Comment (jaseweston) [reply examples/train_model.py]: can't we move this to the world or something? it's annoying people have to put this in their train script isn't it?

File: parlai/agents/memnn_extended/memnn_extended.py [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+
+class MemnnExtendedAgent(Agent):
+    """ Memory Network agent.
+    With setting IM acts as a regular Imitation learning agent
+
+    For more details on IM_feedback, RBI, FP, RBI+FP settings that support 
+    reward-based and forward prediction: https://arxiv.org/abs/1604.06045
+    These models assume that feedback and reward for the current example  
+    immediatly follow the query. 
+    py examples/train_model.py -m memnn_extended -t dbll_babi:task:3_p0.5:f
```
Comment (jaseweston) [inline parlai/agents/memnn_extended/memnn_extended.py]: can ":f" be a little more descriptive?

File: parlai/tasks/dbll_babi/agents.py (hunk 2 above)
Comment (jaseweston) [reply parlai/tasks/dbll_babi/agents.py]: len(self.opt['task'].split(':')) == 4 should more specifically search for a string name and have a comment at least

File: parlai/tasks/dbll_babi/agents.py [hunk 4]
```diff
@@ -55,8 +55,113 @@ def __init__(self, opt, shared=None):
...
+        """Reads data in the fbdialog format.
```
Comment (jaseweston) [inline parlai/tasks/dbll_babi/agents.py]: this data doesnt exist (won't be downloaded, though right (right now)?)

File: parlai/agents/memnn_extended/memnn_extended.py [hunk 5]
```diff
@@ -0,0 +1,492 @@
+# Copyright (c) 2017-present, Facebook, Inc.
```
Comment (jaseweston) [inline parlai/agents/memnn_extended/memnn_extended.py]: maybe call it "feedback" or "fp" or something more specific than "extended" ?

File: parlai/agents/memnn_extended/memnn_extended.py [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+
+import os
+import copy
+import random
+
+from .modules import MemNN, Decoder, to_tensors
+
+
+class MemnnExtendedAgent(Agent):
+    """ Memory Network agent.
```
Comment (jaseweston) [inline parlai/agents/memnn_extended/memnn_extended.py]: slightly more descriptive comment please especially first line

File: examples/train_model.py (hunk 1 above)
Comment (alexholdenmiller) [reply examples/train_model.py]: the multiprocessing docs say:
```
Note that this should be called at most once, and it should be protected inside the if __name__ == '__main__' clause of the main module.
```
it might work in HogwildWorld as well but this motivated me to put it in the main module!

File: parlai/tasks/dbll_babi/agents.py (hunk 2 above)
Comment (alexholdenmiller) [reply parlai/tasks/dbll_babi/agents.py]: I meant the actual data:
that is, you feel good about appending the next x to the current x?

Comment (alexholdenmiller) [reply parlai/tasks/dbll_babi/agents.py]: (I'm understanding that correctly, right?

that the data looks like:
```
Mary moved to the bedroom.
Mary travelled to the garden.
Where is John?
No, that's wrong.
[labels: garden]
```

**Comment by jaseweston** [comment]
if you are understanding i agree it's not great -- but is ok for
experiments. for that reason maybe this should be moved to parlai/projects/
?

File: examples/train_model.py (hunk 1 above)
Comment (ekmb) [reply examples/train_model.py]: I tried putting it in HogwildWorld but was getting errors that the context was already created. Also valid world is also a HogWildWorld but the 'set_start_method('spawn')' call should appear only once in a program.

If in __main__ did you mean something like this:
 try:
        import torch
        from multiprocessing import set_start_method
        if torch.cuda.is_available():
            # CUDA with multiprocessing requires 'spawn' start method
            set_start_method('spawn')
    except:
        pass

File: parlai/tasks/dbll_babi/agents.py (hunk 2 above)
Comment (ekmb) [reply parlai/tasks/dbll_babi/agents.py]: Your understanding is correct.

File: parlai/agents/memnn_extended/memnn_extended.py (hunk 5 above)
Comment (ekmb) [reply parlai/agents/memnn_extended/memnn_extended.py]: It supports fp/rbi/rbi+fp/im so I wanted something more general. I can rename it to 'fp' if you think it won't be confusing.

File: parlai/tasks/dbll_babi/agents.py (hunk 4 above)
Comment (ekmb) [reply parlai/tasks/dbll_babi/agents.py]: No, this method only iterates over the dataset. I think the data is downloaded when a new task is created.

Comment (jaseweston) [reply parlai/tasks/dbll_babi/agents.py]: ok, can we just move this stuff to parlai/projects because it doesn't really fit what we see as a standard ParlAI dataset format. As I said I think it's ok for experiments, but our standard format has to be expecting responses in time sequence format (i.e. you wouldn't get the feedback before you've provided the label!)

**Comment by jaseweston** [comment]
what's happening with this?

**Comment by ekmb** [comment]
I'll fix it in a few weeks, sorry can't work on it now

File: parlai/tasks/dbll_babi/agents.py [hunk 7]
```diff
@@ -55,113 +55,8 @@ def __init__(self, opt, shared=None):
         opt['datafile'] = _path(os.path.join('babi', 'babi1'), params, opt)
```
Comment (jaseweston) [inline parlai/tasks/dbll_babi/agents.py]: so dbll_babi hasnt changed?

Comment (ekmb) [reply parlai/tasks/dbll_babi/agents.py]: No, I reverted the changes I made before
